library/psa_crypto.c

```diff
@@ -6679,7 +6679,9 @@ static psa_status_t psa_pbkdf2_set_salt(psa_pbkdf2_key_derivation_t *pbkdf2,
         return PSA_ERROR_BAD_STATE;
     }
 
-    if (pbkdf2->state == PSA_PBKDF2_STATE_INPUT_COST_SET) {
+    if (data_length == 0) {
+        /* Nothing to do */
```

Comment:
There's a single salt, which can be passed incrementally. The operation in this case is `pbkdf2->salt_length += 0`, not `pbkdf2->salt_length = 0`. Appending a zero-length segment to the salt is a no-op. So that part of the code is right.

That being said, now that I look at this code again, there's another case which isn't right: if `pbkdf2->state == PSA_PBKDF2_STATE_INPUT_SALT_SET && pbkdf2->salt_length == 0`, i.e. for a second salt input after an empty salt input, we're doing `memcpy(dest, malloc(0), 0)` and that's undefined behavior. So I'll fix that as well.

Reply:
> The operation in this case is `pbkdf2->salt_length += 0`, not `pbkdf2->salt_length = 0`.

But when the user firsts starts inputting the salt (i.e. when the state is `PSA_PBKDF2_STATE_INPUT_COST_SET`) we set `pbkdf2->salt_length = data_length` rather than adding to the salt length (see [line 6701](https://github.com/gilles-peskine-arm/mbedtls/blob/73936868b84e19692b05c93f3662a24d1f7a04a1/library/psa_crypto.c#L6701)).

Surely we need to do the same thing even if the first chunk of salt data is zero-length?

Reply:
When the original state is COST_SET, we know that `salt_length` is 0, so `= data_length` and `+= data_length` are equivalent.

Reply:
Where does it get initialised at the start of an operation? I haven't managed to spot it. Is it part of a blanket zeroization of the `operation` struct somewhere?

Reply:
All structures are initialized to zero. Though you now have me worried because unions are involved, and when a union contains structures, this doesn't initialize the whole memory:

```
union {
    int default;
    struct {int one; int two;} bar;
} foo_t;
foo_t = {0};
```

This leaves `foo.bar.two` uninitialized, but in practice I believe all the compilers we test with cause a zero-initialization of the whole structure. Compcert is a known counter-example that doesn't initialize `foo.bar.two`.

Reply:
No, it's fine:
https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.4.0/library/psa_crypto.c#L5919

Reply:
Thanks, I am not so familiar with this area of code!